File: clients/rs/src/features/wrap_feature_interface_invoke.rs
use polywrap_client::{
    builder::{PolywrapClientConfig, PolywrapClientConfigBuilder},
    client::PolywrapClient,
    core::{invoker::Invoker, uri::Uri},
};
use polywrap_client_default_config::SystemClientConfig;
use serde::Serialize;
use serde_json::Value;
use std::error::Error;

#[derive(Serialize)]
struct InnerArg {
    uint8: u8,
    str: String,
}

#[derive(Serialize)]
struct InvokeArgs {
    arg: InnerArg,
}

pub fn run_test_case(_: &Value) -> Result<(), Box<dyn Error>> {
    let root = std::env::current_dir()?.join("../../wraps");
    let interface_uri: Uri = "wrap://authority/interface".try_into().unwrap();
    let binding = root.join("interface-invoke/01-implementation/implementations/as");
    let implementation_path = binding.to_str().unwrap();
    let implementation_uri: Uri = format!("file/{implementation_path}").try_into().unwrap();

    let mut config: PolywrapClientConfig = SystemClientConfig::default().into();
    config.add_interface_implementation(interface_uri, implementation_uri);

    let client: PolywrapClient = PolywrapClient::new(config.into());

    let binding = root.join("interface-invoke/02-wrapper/implementations/as");
    let wrapper_path = binding.to_str().unwrap();
    let wrapper_uri: Uri = format!("fs/{wrapper_path}").try_into().unwrap();

    println!("Invoking moduleMethod");

    let args = polywrap_client::msgpack::to_vec(&InvokeArgs {
        arg: InnerArg {
            uint8: 1,
            str: "Test String 1".to_string(),
        },
    })?;

    let result = client.invoke_raw(&wrapper_uri, "moduleMethod", Some(&args), None, None);

    if let Err(err) = result {
        panic!("{err}")
    }

    println!("Success!");

    Ok(())
}

File: clients/swift/Sources/Readiness/Features/WrapFeatureInterfaceInvoke.swift
import PolywrapClient
import Foundation

public struct Arg: Codable {
    public var uint8: UInt8
    public var str: String

    public init(_ uint8: UInt8, _ str: String) {
        self.uint8 = uint8
        self.str = str
    }
}

public struct ImplementationType: Codable {
    public var arg: Arg
    
    public init(_ arg: Arg) {
        self.arg = arg
    }
}

struct WrapFeatureInterfaceInvokeTest: Feature {
    func runTestCase(input: Any) throws -> Void {
        guard let resourcePath = Bundle.main.resourcePath else {
            fatalError("Resource folder not found")
        }
        let root = resourcePath + "/Readiness_Readiness.bundle/Contents/Resources"
        
        let interfaceUri = try Uri("wrap://ens/interface.eth")
        let implementationUri = try Uri("file/\(root)/wraps/interface-invoke/01-implementation/implementations/as")

        let builder = BuilderConfig()
            .addSystemDefault()
            .addInterfaceImplementation(interfaceUri, implementationUri)

        let client = builder.build()
        print("Invoking moduleMethod")

        let arg = Arg(1, "Test String 1")
        let _: Arg = try client.invoke(uri: implementationUri, method: "moduleMethod", args: ImplementationType(arg))
        print("Success!")
    }
}
